docs/reference/modules/discovery.asciidoc

```diff
@@ -1,30 +1,765 @@
+discovered` which describes the current state of the discovery module.
+
+[float]
+[[modules-discovery-bootstrap-cluster]]
+=== Bootstrapping a cluster
+
+Starting an Elasticsearch cluster for the very first time requires the initial
+set of master-eligible nodes to be explicitly set on one or more of the
+master-eligible nodes in the cluster using this setting:
+
+`cluster.initial_master_nodes`::
+
+    Sets a list of the node names or transport addresses of the initial set of
+    master-eligible nodes in a brand-new cluster. By default this list is
+    empty, meaning that this node expects to join a cluster that has already
+    been bootstrapped.
+
+This setting can be given on the command line when starting up each node, or
+added to the `elasticsearch.yml` configuration file. Once the cluster has
+formed this setting is no longer required and should be removed.
+
+For a cluster with 3 master-eligible nodes (named `master-a`, `master-b` and
+`master-c`) the configuration will look as follows:
+
+[source,yaml]
+--------------------------------------------------
+cluster.initial_master_nodes:
+  - master-a
+  - master-b
+  - master-c
+--------------------------------------------------
+
+Alternatively the IP addresses or hostnames of the nodes can be used. If there
+is more than one Elasticsearch node with the same IP address or hostname then
+the transport ports must also be given
+
+[source,yaml]
+--------------------------------------------------
+cluster.initial_master_nodes:
+  - 10.0.10.101
+  - 10.0.10.102:9300
+  - 10.0.10.102:9301
+  - master-node-hostname
+--------------------------------------------------
+
+It is also possible to set the initial set of master nodes on the command-line
+used to start Elasticsearch:
+
+[source,bash]
+--------------------------------------------------
+$ bin/elasticsearch -Ecluster.initial_master_nodes=master-a,master-b,master-c
+--------------------------------------------------
+
+It is technically sufficient to set this on a single master-eligible node in
+the cluster, and only to mention a single master-eligible node, but this does
+not allow for this single node to fail before the cluster has fully formed. It
+is therefore better to bootstrap using multiple master-eligible nodes. In any
+case, when specifying the list of initial master nodes, **it is vitally
+important** to configure each node with exactly the same list of nodes, to
+prevent two independent clusters from forming. Typically you will set this on
+the nodes that are mentioned in the list of initial master nodes.
+
+NOTE: In alpha releases, all listed master-eligible nodes are required to be
+  discovered before bootstrapping can take place. This requirement will be
+  relaxed in production-ready releases.
+
+WARNING: You must put exactly the same set of initial master nodes in each
+  configuration file in order to be sure that only a single cluster forms during
+  bootstrapping and therefore to avoid the risk of data loss.
```

Comment:
Can we also add a note, that this file is only needed during initial cluster formation. Once the cluster has formed, this file/command line arguments could be removed. If they are not removed, they will be just ignored.

Reply:
I would rather not encourage this behaviour, even if we think it will happen in some setups. It is not wholly safe. I do hear of people using only ephemeral storage for their master nodes (e.g. in Docker) and in that case a full cluster restart could potentially lead to silent data loss if a new cluster were to form and start accepting new data, whereas if the bootstrap configuration is removed after the cluster forms for the first time then the mistake will become obvious. I think "should be removed" is correct, using "should" in the sense of [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt):

    3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
    may exist valid reasons in particular circumstances to ignore a
    particular item, but the full implications must be understood and
    carefully weighed before choosing a different course.

Reply:
Hm, I think we need to discuss it because I currently see no convenient way to do it during, for example, the first k8s deployment. This will require people to prepare two k8s deployments for initial es deployment. The first one to bootstrap the cluster, the second one to restart the cluster without this setting. In this perspective, HTTP bootstrapping API is more appealing.